### PR TITLE
Fix for HHh-17172 to retrieve matching session factory either by uuid *or name* during serialization/deserialization

### PR DESCRIPTION
**Pull request by marceloverdijk**
https://hibernate.atlassian.net/browse/HHH-17172

### CODE REVIEW

**Comment by marceloverdijk** [comment]
Any suggestion how to fix the contribution rules for the conflicting commit messages?

**Comment by Sanne** [comment]
Hi, great work!

> Any suggestion how to fix the contribution rules for the conflicting commit messages?

You should rebase on the latest changes rather than merge; if you rebase now that should make the "merge" commit disappear.
Then squash the two different commits you have to introduce the test ([bc81d3d](https://github.com/hibernate/hibernate-orm/pull/7339/commits/bc81d3dde7de8f5877d51d2c33ea3b91a3288abb) and [690ff23](https://github.com/hibernate/hibernate-orm/pull/7339/commits/690ff23bc70b3c48f99e433da9c6ffab133693b3)) should be one commit)

If you're not familiar with git we can help, but I'd suggest to try first.

File: hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import org.hibernate.Transaction;
+import org.hibernate.cfg.Configuration;
+import org.hibernate.cfg.Environment;
+import org.hibernate.internal.util.SerializationHelper;
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.junit4.BaseCoreFunctionalTestCase;
+import org.junit.Test;
+
+/**
+ * @author Gavin King
```
Comment (Sanne) [inline hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java:24]: Could you either use your own name, or remove the author tags..  they are optional, but if used they should be accurate.

File: hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+	@Override
+	protected String getBaseForMappings() {
+		return "org/hibernate/orm/test/";
+	}
+
+	@Override
+	public void configure(Configuration cfg) {
+		super.configure( cfg );
+		cfg.setProperty( Environment.STATEMENT_BATCH_SIZE, "0" ); // problem on HSQLDB (go figure)
```
Comment (Sanne) [inline hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java:41]: Did you mean with H2 ? We're not testing with HSQLDB

File: hibernate-core/src/main/java/org/hibernate/proxy/map/MapLazyInitializer.java [hunk 3]
```diff
@@ -53,4 +53,9 @@ protected boolean isAllowLoadOutsideTransaction() {
 	protected String getSessionFactoryUuid() {
 		return super.getSessionFactoryUuid();
 	}
+
+	@Override
+	protected String getSessionFactoryName() {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/proxy/map/MapLazyInitializer.java:58]: I might be wrong, but this method seems unnecessary as it's overriding the super method with the same code?
If I misunderstood, please add a comment.

File: hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java [hunk 4]
```diff
@@ -441,6 +445,18 @@ protected String getSessionFactoryUuid() {
 		return sessionFactoryUuid;
 	}
 
+	/**
+	 * Get the session factory name.
+	 *
+	 * This method should only be called during serialization,
+	 * and only makes sense after a call to {@link #prepareForPossibleLoadingOutsideTransaction()}.
+	 *
+	 * @return the session factory UUID.
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/proxy/AbstractLazyInitializer.java:454]:
```suggestion
	 * @return the session factory name.
```

**Review by Sanne — CHANGES_REQUESTED** [review]
Looks promising! Some initial feedback, sorry couldn't review it thouroughly yet.

File: hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java (hunk 1 above)
Comment (marceloverdijk) [reply hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java:24]: Sure. I copied the test class but will adapt it.

File: hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java (hunk 2 above)
Comment (marceloverdijk) [reply hibernate-core/src/test/java/org/hibernate/orm/test/proxy/MultipleSessionFactoriesProxyTest.java:41]: A copy as well; will remove it.

File: hibernate-core/src/main/java/org/hibernate/proxy/map/MapLazyInitializer.java (hunk 3 above)
Comment (marceloverdijk) [reply hibernate-core/src/main/java/org/hibernate/proxy/map/MapLazyInitializer.java:58]: Actually I thought the same initially ;-) but without it, it gives:

```
> Task :hibernate-core:compileJava
hibernate-orm/hibernate-core/src/main/java/org/hibernate/proxy/map/MapProxy.java:123: error: getSessionFactoryName() has protected access in AbstractLazyInitializer
  li.getSessionFactoryName(),    
```

Note that `MapLazyInitializer` also has similar:

```
	@Override
	protected String getSessionFactoryUuid() {
		return super.getSessionFactoryUuid();
	}
```

It needs to be there as `MapProxy` uses them (same package as `MapLazyInitializer`!) but is not extending any of the classed mentioned above but only using `MapLazyInitializer` internally, and hence the latter one needs to expose it.

**Comment by marceloverdijk** [comment]
I've screwed up my local branch.
Will redo the changes (`cherry-pick -n`) and create a new PR with the feedback above resolved as well.

I will close this PR and #7338 (the initial PR with only the test case)

**Comment by marceloverdijk** [comment]
New PR: https://github.com/hibernate/hibernate-orm/pull/7342

**Pull request closed without merging**